pywr/model.py

```diff
@@ -1 +1,154 @@
+        """Add a model."""
+        self.models[name] = model
+        model.parent = self
+
+    @classmethod
+    def loads(cls, data, model=None, path=None, solver=None, **kwargs):
+        """Read JSON data from a string and parse it as a model document"""
+        try:
+            data = json.loads(data)
+        except ValueError as e:
+            message = e.args[0]
+            if path:
+                e.args = ("{} [{}]".format(e.args[0], os.path.basename(path)),)
+            raise (e)
+
+        return cls.load(data, model, path, solver, **kwargs)
+
+    @classmethod
+    def load(cls, data, model=None, path=None, solver=None, **kwargs):
+        if isinstance(data, (str, Path)):
+            # argument is a filename
+            logger.info('Loading model from file: "{}"'.format(path))
+            path = data
+            with open(path, "r") as f:
+                data = f.read()
+            return cls.loads(data, model, path, solver)
+
+        if hasattr(data, "read"):
+            logger.info("Loading model from file-like object.")
+            # argument is a file-like object
+            data = data.read()
+            return cls.loads(data, model, path, solver)
+
+        return cls._load_from_dict(data, model=model, path=path, solver=None, **kwargs)
+
+    @classmethod
+    def _load_from_dict(cls, data, model=None, path=None, solver=None, **kwargs):
+        """Load data from a dictionary."""
+        # data is a dictionary, make a copy to avoid modify the input
+        data = copy.deepcopy(data)
+
+        if model is None:
+            model = cls()
+
+        timestepper_data = data["timestepper"]
+
+        # Load the sub-models
+        sub_model_paths = data["models"]
+        for sub_model_definition in sub_model_paths:
+            sub_model_name = sub_model_definition["name"]
+            sub_model_path = sub_model_definition["path"]
+            sub_model_solver = sub_model_definition.get("solver", solver)
+            if path is not None:
+                sub_model_path = os.path.join(os.path.dirname(path), sub_model_path)
+
+            sub_model = Model.load(sub_model_path, path=path, solver=sub_model_solver)
+            sub_model.timestepper.start = timestepper_data["start"]
+            sub_model.timestepper.end = timestepper_data["end"]
+            sub_model.timestepper.timestep = timestepper_data["timestep"]
+            model.add_model(sub_model_name, sub_model)
+
+        return model
+
+    @property
+    def dirty(self) -> bool:
+        """Returns true if any of the models are 'dirty'."""
+        return any(
+            model.dirty or model.timestepper.dirty for model in self.models.values()
+        )
+
+    def setup(self):
+        for model in self.models.values():
+            model.setup()
+
+    def reset(self):
+        for model in self.models.values():
+            model.reset()
+
+    # def _step(self):
+    #     for model in self.models.values():
```

Comment:
You thinking of keeping this implementation? Might be useful to be able to specify this one or the one below in the model definition. Though I guess it would change how the new parameters work.

Reply:
No, removed.